### ISSUE TITLE
How to properly clip gradients with data parallel training?

### ISSUE THREAD

**Issue by myleott** [pytorch/xla#1708]
The XLA API recommends that we call `xm.optimizer_step(optimizer)`, which [allreduces the gradients and performs the optimization step](https://github.com/pytorch/xla/blob/ffde50813f01b57d6e782a63aac6453bfa12ffdf/torch_xla/core/xla_model.py#L429-L455).

Gradient clipping is typically performed on the reduced gradients, however it seems the current API doesn't give us access to the reduced gradients before taking the optimization step.

The code for `optimizer_step` is pretty simple, so I'm also wondering why we have `xm.optimizer_step` interface in the first place, compared to the more standard DistributedDataParallel interface that PyTorch uses. For example, in fairseq we have a [very simple DistributedDataParallel wrapper](https://github.com/pytorch/fairseq/blob/master/fairseq/legacy_distributed_data_parallel.py) that just calls allreduce, so it seems something like that could be a drop-in replacement here.

Is there something else special happening with `xm.optimizer_step` that I'm missing?

**Comment by dlibenzi** [comment]
There was discussion back then about whether to change/hook into each optimizer code, or have a separate API.
We need to hook into the gradients before the optimizer uses them, in order to inject `AllReduce(SUM)` in them.
The reason of doing that is that there will be a single, fully fused, TPU computation, instead of two separate ones.
This saves time and memory, as certain intermediate result will not be materialized at all.
We have an `xm.all_reduce()` API you can use if you want to do something different, but we will have to provide another version of our `xm.optimizer_step()` to make sure that things we want to run there, still do.
If you want you can even run the vanilla optimizer step and then reduce the weights 😄

**Comment by dlibenzi** [comment]
With the new API one should be able to do:
```Python
xm.reduce_gradients(optimizer)
clip(optimizer)
optimizer.step()
```

https://github.com/pytorch/xla/blob/3c94753b8e6b3778b9f31fab8e78453f9c8e44dd/torch_xla/core/xla_model.py#L424

**Comment by TheFaheem** [comment]
> 

can't we pass grad clip value?
